src/core/observer/watcher.js

```diff
@@ -230,30 +230,29 @@ export default class Watcher {
  * is collected as a "deep" dependency.
  */
 const seenObjects = new Set()
-function traverse (val: any, seen?: Set) {
+function traverse (val: any) {
+  seenObjects.clear()
+  _traverse(val, seenObjects)
+}
+
+function _traverse (val: any, seen: Set) {
   let i, keys
-  if (!seen) {
-    seen = seenObjects
-    seen.clear()
-  }
   const isA = Array.isArray(val)
-  const isO = isObject(val)
-  if ((isA || isO) && Object.isExtensible(val)) {
+  if ((isA || isObject(val)) && Object.isExtensible(val)) {
```

Comment:
Would you mind use a faster return pattern? That is, if `!isA && !isObject(val)` then `return`.

Because the following code has assumption that val must be either Array or Object. A `return` can indicate this better.(also reduce one indentation level)